README.rst
===============================
AddressBook
===============================

.. image:: https://img.shields.io/travis/BhattiMarry/addressbook.svg
        :target: https://travis-ci.org/BhattiMarry/addressbook

.. image:: https://img.shields.io/pypi/v/addressbook.svg
        :target: https://pypi.python.org/pypi/addressbook


Simple addressbook api for python.

* Free software: Apache 2.0
* Documentation: (COMING SOON!) https://addressbook.readthedocs.org.

Features
--------

* Add people's group to the address book
* Add persons to the address book
* Search and dump persons' info


HISTORY.rst
.. :changelog:

History
-------

Version: 1.0

Date: Sept 24, 2016

Commitor: Umair Bhatti

Detail: Initial Release

----------------------------------------------------------------------------------------------------------------------------------


tests/tests.py
import unittest
from AddressBook import AddressBook
from AddressBookHandler import AddressBookHandler

class AddressBookTests(unittest.TestCase):
    def setUp(self):
        self.obj = AddressBookHandler()

    def test_address_book_object(self):
        assert not AddressBook("Umair", "Bhatti", "GHQ", "marry@smart.com", "5001676", "Human") == None

    def test_add_group(self):
        self.obj.add_group('Human')

    def test_add_person(self):
        self.obj.add_person("Jawad", "Aslam", "PDK", "jawad@hafiz.com", "5519726", "Human")
        assert self.obj.count_persons() > 0

        self.obj.add_person("Umair", "Bhatti", "GHQ", "marry@smart.com", "5001676", "Human")
        assert self.obj.count_persons() == 2

    def test_get_address_book(self):
        assert type(self.obj.get_address_book()) == list

    def test_ptrint_short_info(self):
        allPersons = self.obj.get_address_book()
        if len(allPersons) > 0:
            self.obj.print_short_info(allPersons[0])

    def test_get_persons_in_group(self):
        allGroups = self.obj.get_all_groups()
        if len(allGroups) > 0:
            self.obj.get_persons_in_group(self.obj.get_all_groups()[0])
        else:
            self.obj.get_persons_in_group("Default")

    def test_print_person_info(self):
        allPersons = self.obj.get_address_book()
        if len(allPersons) > 0:
            self.obj.print_person_info(allPersons[0])

    def test_get_group_info_of_person(self):
        allPersons = self.obj.get_address_book()
        if len(allPersons) > 0:
            self.obj.get_group_info_of_person(allPersons[0].fname)

    def test_get_person_info_by_name(self):
        allPersons = self.obj.get_address_book()
        if len(allPersons) > 0:
            self.obj.get_person_info_by_name(allPersons[0].fname)

    def test_get_person_info_by_email(self):
        allPersons = self.obj.get_address_book()
        if len(allPersons) > 0:
            self.obj.get_person_info_by_name(allPersons[0].email)

    def tearDown(self):
        self.obj = None

if __name__ == '__main__':
    unittest.main()


addressbook/_version.py

# This file helps to compute a version number in source trees obtained from
# git-archive tarball (such as those provided by githubs download-from-tag
# feature). Distribution tarballs (built by setup.py sdist) and build
# directories (produced by setup.py build) will contain a much shorter file
# that just contains the computed version number.

# This file is released into the public domain. Generated by
# versioneer-0.15 (https://github.com/warner/python-versioneer)

import errno
import os
import re
import subprocess
import sys


def get_keywords():
    # these strings will be replaced by git during git-archive.
    # setup.py/versioneer.py will grep for the variable names, so they must
    # each be defined on a line of their own. _version.py will just call
    # get_keywords().
    git_refnames = "$Format:%d$"
    git_full = "$Format:%H$"
    keywords = {"refnames": git_refnames, "full": git_full}
    return keywords


class VersioneerConfig:
    pass


def get_config():
    # these strings are filled in when 'setup.py versioneer' creates
    # _version.py
    cfg = VersioneerConfig()
    cfg.VCS = "git"
    cfg.style = "pep440"
    cfg.tag_prefix = ""
    cfg.parentdir_prefix = "example-"
    cfg.versionfile_source = "example/_version.py"
    cfg.verbose = False
    return cfg


class NotThisMethod(Exception):
    pass


LONG_VERSION_PY = {}
HANDLERS = {}


def register_vcs_handler(vcs, method):  # decorator
    def decorate(f):
        if vcs not in HANDLERS:
            HANDLERS[vcs] = {}
        HANDLERS[vcs][method] = f
        return f
    return decorate


def run_command(commands, args, cwd=None, verbose=False, hide_stderr=False):
    assert isinstance(commands, list)
    p = None
    for c in commands:
        try:
            dispcmd = str([c] + args)
            # remember shell=False, so use git.cmd on windows, not just git
            p = subprocess.Popen([c] + args, cwd=cwd, stdout=subprocess.PIPE,
                                 stderr=(subprocess.PIPE if hide_stderr
                                         else None))
            break
        except EnvironmentError:
            e = sys.exc_info()[1]
            if e.errno == errno.ENOENT:
                continue
            if verbose:
                print("unable to run %s" % dispcmd)
                print(e)
            return None
    else:
        if verbose:
            print("unable to find command, tried %s" % (commands,))
        return None
    stdout = p.communicate()[0].strip()
    if sys.version_info[0] >= 3:
        stdout = stdout.decode()
    if p.returncode != 0:
        if verbose:
            print("unable to run %s (error)" % dispcmd)
        return None
    return stdout


def versions_from_parentdir(parentdir_prefix, root, verbose):
    # Source tarballs conventionally unpack into a directory that includes
    # both the project name and a version string.
    dirname = os.path.basename(root)
    if not dirname.startswith(parentdir_prefix):
        if verbose:
            print("guessing rootdir is '%s', but '%s' doesn't start with "
                  "prefix '%s'" % (root, dirname, parentdir_prefix))
        raise NotThisMethod("rootdir doesn't start with parentdir_prefix")
    return {"version": dirname[len(parentdir_prefix):],
            "full-revisionid": None,
            "dirty": False, "error": None}


@register_vcs_handler("git", "get_keywords")
def git_get_keywords(versionfile_abs):
    # the code embedded in _version.py can just fetch the value of these
    # keywords. When used from setup.py, we don't want to import _version.py,
    # so we do it with a regexp instead. This function is not used from
    # _version.py.
    keywords = {}
    try:
        f = open(versionfile_abs, "r")
        for line in f.readlines():
            if line.strip().startswith("git_refnames ="):
                mo = re.search(r'=\s*"(.*)"', line)
                if mo:
                    keywords["refnames"] = mo.group(1)
            if line.strip().startswith("git_full ="):
                mo = re.search(r'=\s*"(.*)"', line)
                if mo:
                    keywords["full"] = mo.group(1)
        f.close()
    except EnvironmentError:
        pass
    return keywords


@register_vcs_handler("git", "keywords")
def git_versions_from_keywords(keywords, tag_prefix, verbose):
    if not keywords:
        raise NotThisMethod("no keywords at all, weird")
    refnames = keywords["refnames"].strip()
    if refnames.startswith("$Format"):
        if verbose:
            print("keywords are unexpanded, not using")
        raise NotThisMethod("unexpanded keywords, not a git-archive tarball")
    refs = set([r.strip() for r in refnames.strip("()").split(",")])
    # starting in git-1.8.3, tags are listed as "tag: foo-1.0" instead of
    # just "foo-1.0". If we see a "tag: " prefix, prefer those.
    TAG = "tag: "
    tags = set([r[len(TAG):] for r in refs if r.startswith(TAG)])
    if not tags:
        # Either we're using git < 1.8.3, or there really are no tags. We use
        # a heuristic: assume all version tags have a digit. The old git %d
        # expansion behaves like git log --decorate=short and strips out the
        # refs/heads/ and refs/tags/ prefixes that would let us distinguish
        # between branches and tags. By ignoring refnames without digits, we
        # filter out many common branch names like "release" and
        # "stabilization", as well as "HEAD" and "master".
        tags = set([r for r in refs if re.search(r'\d', r)])
        if verbose:
            print("discarding '%s', no digits" % ",".join(refs-tags))
    if verbose:
        print("likely tags: %s" % ",".join(sorted(tags)))
    for ref in sorted(tags):
        # sorting will prefer e.g. "2.0" over "2.0rc1"
        if ref.startswith(tag_prefix):
            r = ref[len(tag_prefix):]
            if verbose:
                print("picking %s" % r)
            return {"version": r,
                    "full-revisionid": keywords["full"].strip(),
                    "dirty": False, "error": None
                    }
    # no suitable tags, so version is "0+unknown", but full hex is still there
    if verbose:
        print("no suitable tags, using unknown + full revision id")
    return {"version": "0+unknown",
            "full-revisionid": keywords["full"].strip(),
            "dirty": False, "error": "no suitable tags"}


@register_vcs_handler("git", "pieces_from_vcs")
def git_pieces_from_vcs(tag_prefix, root, verbose, run_command=run_command):
    # this runs 'git' from the root of the source tree. This only gets called
    # if the git-archive 'subst' keywords were *not* expanded, and
    # _version.py hasn't already been rewritten with a short version string,
    # meaning we're inside a checked out source tree.

    if not os.path.exists(os.path.join(root, ".git")):
        if verbose:
            print("no .git in %s" % root)
        raise NotThisMethod("no .git directory")

    GITS = ["git"]
    if sys.platform == "win32":
        GITS = ["git.cmd", "git.exe"]
    # if there is a tag, this yields TAG-NUM-gHEX[-dirty]
    # if there are no tags, this yields HEX[-dirty] (no NUM)
    describe_out = run_command(GITS, ["describe", "--tags", "--dirty",
                                      "--always", "--long"],
                               cwd=root)
    # --long was added in git-1.5.5
    if describe_out is None:
        raise NotThisMethod("'git describe' failed")
    describe_out = describe_out.strip()
    full_out = run_command(GITS, ["rev-parse", "HEAD"], cwd=root)
    if full_out is None:
        raise NotThisMethod("'git rev-parse' failed")
    full_out = full_out.strip()

    pieces = {}
    pieces["long"] = full_out
    pieces["short"] = full_out[:7]  # maybe improved later
    pieces["error"] = None

    # parse describe_out. It will be like TAG-NUM-gHEX[-dirty] or HEX[-dirty]
    # TAG might have hyphens.
    git_describe = describe_out

    # look for -dirty suffix
    dirty = git_describe.endswith("-dirty")
    pieces["dirty"] = dirty
    if dirty:
        git_describe = git_describe[:git_describe.rindex("-dirty")]

    # now we have TAG-NUM-gHEX or HEX

    if "-" in git_describe:
        # TAG-NUM-gHEX
        mo = re.search(r'^(.+)-(\d+)-g([0-9a-f]+)$', git_describe)
        if not mo:
            # unparseable. Maybe git-describe is misbehaving?
            pieces["error"] = ("unable to parse git-describe output: '%s'"
                               % describe_out)
            return pieces

        # tag
        full_tag = mo.group(1)
        if not full_tag.startswith(tag_prefix):
            if verbose:
                fmt = "tag '%s' doesn't start with prefix '%s'"
                print(fmt % (full_tag, tag_prefix))
            pieces["error"] = ("tag '%s' doesn't start with prefix '%s'"
                               % (full_tag, tag_prefix))
            return pieces
        pieces["closest-tag"] = full_tag[len(tag_prefix):]

        # distance: number of commits since tag
        pieces["distance"] = int(mo.group(2))

        # commit: short hex revision ID
        pieces["short"] = mo.group(3)

    else:
        # HEX: no tags
        pieces["closest-tag"] = None
        count_out = run_command(GITS, ["rev-list", "HEAD", "--count"],
                                cwd=root)
        pieces["distance"] = int(count_out)  # total number of commits

    return pieces


def plus_or_dot(pieces):
    if "+" in pieces.get("closest-tag", ""):
        return "."
    return "+"


def render_pep440(pieces):
    # now build up version string, with post-release "local version
    # identifier". Our goal: TAG[+DISTANCE.gHEX[.dirty]] . Note that if you
    # get a tagged build and then dirty it, you'll get TAG+0.gHEX.dirty

    # exceptions:
    # 1: no tags. git_describe was just HEX. 0+untagged.DISTANCE.gHEX[.dirty]

    if pieces["closest-tag"]:
        rendered = pieces["closest-tag"]
        if pieces["distance"] or pieces["dirty"]:
            rendered += plus_or_dot(pieces)
            rendered += "%d.g%s" % (pieces["distance"], pieces["short"])
            if pieces["dirty"]:
                rendered += ".dirty"
    else:
        # exception #1
        rendered = "0+untagged.%d.g%s" % (pieces["distance"],
                                          pieces["short"])
        if pieces["dirty"]:
            rendered += ".dirty"
    return rendered


def render_pep440_pre(pieces):
    # TAG[.post.devDISTANCE] . No -dirty

    # exceptions:
    # 1: no tags. 0.post.devDISTANCE

    if pieces["closest-tag"]:
        rendered = pieces["closest-tag"]
        if pieces["distance"]:
            rendered += ".post.dev%d" % pieces["distance"]
    else:
        # exception #1
        rendered = "0.post.dev%d" % pieces["distance"]
    return rendered


def render_pep440_post(pieces):
    # TAG[.postDISTANCE[.dev0]+gHEX] . The ".dev0" means dirty. Note that
    # .dev0 sorts backwards (a dirty tree will appear "older" than the
    # corresponding clean one), but you shouldn't be releasing software with
    # -dirty anyways.

    # exceptions:
    # 1: no tags. 0.postDISTANCE[.dev0]

    if pieces["closest-tag"]:
        rendered = pieces["closest-tag"]
        if pieces["distance"] or pieces["dirty"]:
            rendered += ".post%d" % pieces["distance"]
            if pieces["dirty"]:
                rendered += ".dev0"
            rendered += plus_or_dot(pieces)
            rendered += "g%s" % pieces["short"]
    else:
        # exception #1
        rendered = "0.post%d" % pieces["distance"]
        if pieces["dirty"]:
            rendered += ".dev0"
        rendered += "+g%s" % pieces["short"]
    return rendered


def render_pep440_old(pieces):
    # TAG[.postDISTANCE[.dev0]] . The ".dev0" means dirty.

    # exceptions:
    # 1: no tags. 0.postDISTANCE[.dev0]

    if pieces["closest-tag"]:
        rendered = pieces["closest-tag"]
        if pieces["distance"] or pieces["dirty"]:
            rendered += ".post%d" % pieces["distance"]
            if pieces["dirty"]:
                rendered += ".dev0"
    else:
        # exception #1
        rendered = "0.post%d" % pieces["distance"]
        if pieces["dirty"]:
            rendered += ".dev0"
    return rendered


def render_git_describe(pieces):
    # TAG[-DISTANCE-gHEX][-dirty], like 'git describe --tags --dirty
    # --always'

    # exceptions:
    # 1: no tags. HEX[-dirty]  (note: no 'g' prefix)

    if pieces["closest-tag"]:
        rendered = pieces["closest-tag"]
        if pieces["distance"]:
            rendered += "-%d-g%s" % (pieces["distance"], pieces["short"])
    else:
        # exception #1
        rendered = pieces["short"]
    if pieces["dirty"]:
        rendered += "-dirty"
    return rendered


def render_git_describe_long(pieces):
    # TAG-DISTANCE-gHEX[-dirty], like 'git describe --tags --dirty
    # --always -long'. The distance/hash is unconditional.

    # exceptions:
    # 1: no tags. HEX[-dirty]  (note: no 'g' prefix)

    if pieces["closest-tag"]:
        rendered = pieces["closest-tag"]
        rendered += "-%d-g%s" % (pieces["distance"], pieces["short"])
    else:
        # exception #1
        rendered = pieces["short"]
    if pieces["dirty"]:
        rendered += "-dirty"
    return rendered


def render(pieces, style):
    if pieces["error"]:
        return {"version": "unknown",
                "full-revisionid": pieces.get("long"),
                "dirty": None,
                "error": pieces["error"]}

    if not style or style == "default":
        style = "pep440"  # the default

    if style == "pep440":
        rendered = render_pep440(pieces)
    elif style == "pep440-pre":
        rendered = render_pep440_pre(pieces)
    elif style == "pep440-post":
        rendered = render_pep440_post(pieces)
    elif style == "pep440-old":
        rendered = render_pep440_old(pieces)
    elif style == "git-describe":
        rendered = render_git_describe(pieces)
    elif style == "git-describe-long":
        rendered = render_git_describe_long(pieces)
    else:
        raise ValueError("unknown style '%s'" % style)

    return {"version": rendered, "full-revisionid": pieces["long"],
            "dirty": pieces["dirty"], "error": None}


def get_versions():
    # I am in _version.py, which lives at ROOT/VERSIONFILE_SOURCE. If we have
    # __file__, we can work backwards from there to the root. Some
    # py2exe/bbfreeze/non-CPython implementations don't do __file__, in which
    # case we can only use expanded keywords.

    cfg = get_config()
    verbose = cfg.verbose

    try:
        return git_versions_from_keywords(get_keywords(), cfg.tag_prefix,
                                          verbose)
    except NotThisMethod:
        pass

    try:
        root = os.path.realpath(__file__)
        # versionfile_source is the relative path from the top of the source
        # tree (where the .git directory might live) to this file. Invert
        # this to find the root from __file__.
        for i in cfg.versionfile_source.split('/'):
            root = os.path.dirname(root)
    except NameError:
        return {"version": "0+unknown", "full-revisionid": None,
                "dirty": None,
                "error": "unable to find root of source tree"}

    try:
        pieces = git_pieces_from_vcs(cfg.tag_prefix, root, verbose)
        return render(pieces, cfg.style)
    except NotThisMethod:
        pass

    try:
        if cfg.parentdir_prefix:
            return versions_from_parentdir(cfg.parentdir_prefix, root, verbose)
    except NotThisMethod:
        pass

    return {"version": "0+unknown", "full-revisionid": None,
            "dirty": None,
            "error": "unable to compute version"}


addressbook/AddressBook.py
# -*- coding: utf-8 -*-
'''

AddressBook class.

This module provides the Address Book object containing very primary information of a person.

:copyright: (c) 2016 Umair Bhatti
:licence: Apache2, see LICENCE file for more details.

'''
class AddressBook(object):

    def __init__(self, FirstName, LastName, StreetAddress, Email, Phone, Group):
        self.fname = FirstName
        self.lname = LastName
        if StreetAddress:
            self.stAddress = StreetAddress
        else:
            self.stAddress = "N/A"

        if Email:
            self.email = Email
        else:
            self.email = "N/A"

        if Phone:
            self.phone = Phone
        else:
            self.phone = "N/A"

        if Group:
            self.group = Group
        else:
            self.group = "N/A"


addressbook/AddressBookHandler.py
# -*- coding: utf-8 -*-

'''

AddressBookHandler class.

This module provides the implementation of Address Book using AddressBook objects.

:copyright: (c) 2016 Umair Bhatti
:licence: Apache2, see LICENCE file for more details.

'''

from AddressBook import AddressBook

class AddressBookHandler():
    addressBook = []
    groups = []

    def __init__(self):
        addressBook = []
        groups = []

    def add_group(self, Group_Name):
        if Group_Name:
            AddressBookHandler.groups.append(Group_Name)

    def check_sublist(self, list1, list2):
        for item in list1:
            if item not in list2:
                return False
        return True

    def add_person(self, FirstName, LastName, StreetAddress, Email, Phone, Group):
        if Group:
            if isinstance(Group, list):
                if not self.check_sublist(Group, AddressBookHandler.groups):
                    print("Group name not found in group list. Add the group first.")
                    return
            else:
                if Group not in AddressBookHandler.groups:
                    print("Group name not found in group list. Add the group first.")
                    return
            
        obj = AddressBook(FirstName, LastName, StreetAddress, Email, Phone, Group)
        AddressBookHandler.addressBook.append(obj)

    def count_persons(self):
        len1 = len(AddressBookHandler.addressBook)
        print len1
        return len1

    def get_address_book(self):
        return AddressBookHandler.addressBook

    def clear_address_book(self):
        print 'Clearing address book entries...'
        AddressBookHandler.addressBook = []

    def clear_address_book_groups(self):
        print 'Clearing address book groups...'
        AddressBookHandler.groups = []

    def print_short_info(self, AddressBook):
        print "First Name: %s, Last Name: %s, Group: %s" %(AddressBook.fname, AddressBook.lname, AddressBook.group)

    def get_all_groups(self):
    	all_groups = AddressBookHandler.groups
    	print all_groups
        return all_groups

    def get_persons_in_group(self, GroupName):
        members = []
        if not GroupName:
            print "Error: `No group name specified.`"
            return
        all_groups = self.get_all_groups()

        if not all_groups and GroupName not in all_groups:
            print 'Error: `No group(s) registered with this name.`'
            return

        if not AddressBookHandler.addressBook:
            print "Address book is empty. Add some contacts first."
            return

        addr_book = self.get_address_book()
        for i in xrange(0, len(addr_book)):
            if GroupName in addr_book[i].group:
                self.print_short_info(addr_book[i])
                members.append(addr_book[i])

        if members:
            return members
        else:
            print 'No person in this group.'

    def print_person_info(self, AddressBook):
        print 'First Name       :   %s' %AddressBook.fname
        print 'Last Name        :   %s' %AddressBook.lname
        print 'Street Address   :   %s' %AddressBook.stAddress
        print 'Email Address    :   %s' %AddressBook.email
        print 'Phone Number     :   %s' %AddressBook.phone
        print 'Address Group    :   %s' %AddressBook.group

    def get_group_info_of_person(self, _fname):
        nameFlag = True
        groups = []
        addr_book = self.get_address_book()
        for i in xrange(0, len(addr_book)):
            if addr_book[i].fname == _fname:
                nameFlag = False
                self.print_short_info(addr_book[i])
                groups.append(addr_book[i].group)
        if nameFlag:
            print 'Error: `Name not found in the address book.`'
        else:
            return groups

    def get_person_info_by_name(self, Name):
        members = []
        _fname = None
        _lname = None
        if not Name:
            print 'Error: `No name given.`'
            return
        nameTokens = Name.split(' ')
        if len(nameTokens) > 1:
            _fname = (['', ' '][len(nameTokens) > 2]).join(nameTokens[:len(nameTokens) - 1])
            _lname = ''.join(nameTokens[len(nameTokens) - 1])

        addr_book = self.get_address_book()
        for i in xrange(0, len(addr_book)):
            if _fname and _lname:
                if _fname == addr_book[i].fname and _lname == addr_book[i].lname:
                    members.append(addr_book[i])
                    self.print_short_info(addr_book[i])
            else:
                if Name in [addr_book[i].fname, addr_book[i].lname]:
                    members.append(addr_book[i])
                    self.print_short_info(addr_book[i])
        if members:
            return members
        else:
            print 'Name not found in the address book.'

    def get_person_info_by_email(self, Email):
        eName = None
        members = []
        if not Email:
            print 'Error: `No email address provided to look for.`'
            return
        addr_book = self.get_address_book()
        for i in range(0, len(addr_book)):
            if isinstance(addr_book[i].email, str):
                if Email in addr_book[i].email:
                    self.print_short_info(addr_book[i])
                    members.append(addr_book[i])
            elif isinstance(addr_book[i].email, list):
                for email in addr_book[i].email:
                    if Email in item:
                        self.print_short_info(addr_book[i])
                        members.append(addr_book[i])
        if members:
            return members
        else:
            print 'No person with given email address.'


sample/sample.py
from AddressBookHandler import AddressBookHandler
obj = AddressBookHandler()

#adds a person group with the given name.
obj.add_group('Human')

#adds a person with given details.
obj.add_person('Umair', 'Bhatti', 'KHI', 'umair@marry.com', '5001676', 'Human')

obj.add_group('Alien')

obj.add_person('Jawad', 'Aslam', 'ISB', 'hafiz@jawad.com', '5519726', 'Alien')
obj.add_person('Saim', 'Ali', 'PDK', 'saim@don.com', '9872208', ['Human', 'Alien'])
obj.add_person('Ch', 'Uzair', 'RWP', 'malik@uzair.com', '5028140', 'Malik')

obj.add_group('Malik')

obj.add_person('Ch', 'Uzair', 'RWP', 'malik@uzair.com', '5028140', 'Malik')

#returns and prints the total number of persons in the address book.
obj.count_persons()

#returns and prints a list of groups in address book
obj.get_all_groups()

#returns and prints the list of person(s) matching with the given group name
obj.get_persons_in_group('Malik')
obj.get_persons_in_group('Alien')
obj.get_persons_in_group('Human')

#returns and prints a list of group(s) against a given first name
obj.get_group_info_of_person('Umair')

#returns and prints a list of person(s) matching with given name
obj.get_person_info_by_name('Ch')

#returns and prints a list of person matching the email given
person = obj.get_person_info_by_email('don.com')

obj.print_person_info(person[0])
